usr/share/rear/rescue/default/900_clone_users_and_groups.sh

```diff
@@ -1,41 +1,60 @@
+# Copy users and groups to the ReaR recovery system:
+
+if is_true "$CLONE_ALL_USERS_GROUPS" ; then
+    # In particular exclude a possible NIS extension line '+::::::'
+    CLONE_USERS=( $( grep -o '^[[:alnum:]_][^:]*' /etc/passwd ) )
+    # In particular exclude a possible NIS extension line '+:::'
+    CLONE_GROUPS=( $( grep -o '^[[:alnum:]_][^:]*' /etc/group ) )
 fi
 
-Log "Cloning users: ${CLONE_USERS[@]}"
-for u in "${CLONE_USERS[@]}" ; do
-	# go over all users
-	if pwd=$(getent passwd "$u") ; then
-		# pwd="daemon:x:2:2:Daemon:/sbin:/bin/bash"
-		# if the user exists, add to the passwd in rescue system
-		# skip if this user exists already in the rescue system
-		grep -q "^$pwd:" $ROOTFS_DIR/etc/passwd && continue
-		echo "$pwd" >>$ROOTFS_DIR/etc/passwd
-		# strip gid from passwd line
-		pwd="${pwd#*:*:*:}"
-		gid=${pwd%%:*}
-		# add gid to groups to collect
-		CLONE_GROUPS=( "${CLONE_GROUPS[@]}" "$gid" )
-	else
-		Debug "WARNING: Could not collect user info for '$u'"
-	fi
-done
+local user=""
+local passwd_entry=""
+local groupID=""
+local group_entry=""
+local group=""
 
-Log "Cloning groups: ${CLONE_GROUPS[@]}"
-for g in "${CLONE_GROUPS[@]}" ; do
-	# go over all users
-	if grp=$(getent group "$g") ; then
-		# grp="daemon:x:2:"
-		# if the group  exists, add to the group in rescue system
-		# skip if this user exists already in the rescue system
-		grep -q "^$grp" $ROOTFS_DIR/etc/group && continue
-		echo "$grp" >>$ROOTFS_DIR/etc/group
-	else
-		Debug "WARNING: Could not collect group info for '$g'"
-	fi
+# For the 'test' one must have all array members as a single word i.e. "${name[*]}"
+# because it should succeed when there is any non-empty array member, not necessarily the first one:
+test "${CLONE_USERS[*]}" && Log "Cloning users: ${CLONE_USERS[@]}"
+for user in "${CLONE_USERS[@]}" ; do
+    # Skip if the user exists already in the ReaR recovery system:
+    grep "^$user:" $ROOTFS_DIR/etc/passwd 1>&2 && continue
+    # Skip if the user does not exist in the current system:
+    if ! passwd_entry="$( getent passwd $user )" ; then
+        Debug "Cannot clone $user because it does not exist"
+        continue
+    fi
+    # Prepare to also add the user's group to the CLONE_GROUPS array:
+    # passwd_entry="user:password:UID:GID:description:HOMEdirectory:shell
+    groupID="$( echo "$passwd_entry" | cut -d ':' -f '4' )"
+    if ! group_entry="$( getent group $groupID )" ; then
+        Debug "Cannot clone $user because its group $groupID does not exist"
+        continue
+    fi
+    # Add the user to /etc/passwd in the ReaR recovery system:
+    echo "$passwd_entry" >>$ROOTFS_DIR/etc/passwd
+    # Add the user's group to the CLONE_GROUPS array (unless already there):
+    # group_entry="group:passwd:GID:userlist"
+    group="$( echo "$group_entry" | cut -d ':' -f '1' )"
+    # Skip if the group is already in the CLONE_GROUPS array:
+    IsInArray "$group" "${CLONE_GROUPS[@]}" && continue
```

Comment:
I'd prefer here 
```bash
if !  IsInArray "$group" "${CLONE_GROUPS[@]}" ; then
    CLONE_GROUPS=( "${CLONE_GROUPS[@]}" "$group" )
fi
```
to make it easier to read. I would use `continue` only to denote skipping the entire while block for a certain item.

Reply:
I prefer "return early return often", cf.
https://github.com/rear/rear/wiki/Coding-Style

Reply:
For me "continue" means "continue with the next item"
and that is exactly what it should do in this case.

Reply:
Yes, it does that. However, the next developer who will add more stuff below will be easily fooled by that. That's why I find the use of continue in this case not optimum. With the if clause it is clear that the array check is related to the array addition. A developer can add code before or after that without any worries.

In other words: I prefer locally understandable logic over logic that jumps far away. I agree with the coding style piece you refer to but I don't agree that the generalization for while loops is a conclusion from that.